File: s/sum-of-root-to-leaf-binary-numbers.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  int ans;
  void dfs(TreeNode* node, int cur) {
    int nval = cur + node->val;
    if (node->left == nullptr && node->right == nullptr) {
      ans += nval;
      return;
    }
    if (node->left != nullptr) dfs(node->left, nval * 2);
    if (node->right != nullptr) dfs(node->right, nval * 2);
  }
public:
  int sumRootToLeaf(TreeNode* root) {
    if (!root) return 0;
    ans = 0;
    dfs(root, 0);
    return ans;
  }
};

File: s/sum-of-root-to-leaf-binary-numbers.erl
%% Definition for a binary tree node.
%%
%% -record(tree_node, {val = 0 :: integer(),
%%                     left = null  :: 'null' | #tree_node{},
%%                     right = null :: 'null' | #tree_node{}}).

-spec sum_root_to_leaf(Root :: #tree_node{} | null) -> integer().
sum_root_to_leaf(Root) ->
  case Root of
    null -> 0;
    _ -> sum_root_to_leaf(<<>>, Root)
  end.

sum_root_to_leaf(Bits, #tree_node{val = Val, left = Left, right = Right}) ->
  NextBits = <<Bits/binary, (Val + $0)>>,
  case {Left, Right} of
    {null, null} ->
      binary_to_integer(NextBits, 2);
    _ ->
      LeftSum =
        case Left of
          null -> 0;
          _ -> sum_root_to_leaf(NextBits, Left)
        end,
      RightSum =
        case Right of
          null -> 0;
          _ -> sum_root_to_leaf(NextBits, Right)
        end,
      LeftSum + RightSum
  end.
